src/cdk/tree/tree.ts

```diff
@@ -133,12 +220,56 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
     end: Number.MAX_VALUE,
   });
 
-  constructor(private _differs: IterableDiffers, private _changeDetectorRef: ChangeDetectorRef) {}
+  /** Keep track of which nodes are expanded. */
+  private _expansionModel?: SelectionModel<K>;
+
+  /**
+   * Maintain a synchronous cache of flattened data nodes. This will only be
+   * populated after initial render, and in certain cases, will be delayed due to
+   * relying on Observable `getChildren` calls.
+   */
+  private _flattenedNodes: BehaviorSubject<readonly T[]> = new BehaviorSubject<readonly T[]>([]);
+
+  /** The automatically determined node type for the tree. */
+  private _nodeType: BehaviorSubject<'flat' | 'nested' | null> = new BehaviorSubject<
+    'flat' | 'nested' | null
+  >(null);
+
+  /** The mapping between data and the node that is rendered. */
+  private _nodes: BehaviorSubject<Map<K, CdkTreeNode<T, K>>> = new BehaviorSubject(
+    new Map<K, CdkTreeNode<T, K>>(),
+  );
+
+  /**
+   * Synchronous cache of nodes for the `TreeKeyManager`. This is separate
+   * from `_flattenedNodes` so they can be independently updated at different
+   * times.
+   */
+  private _keyManagerNodes: BehaviorSubject<readonly T[]> = new BehaviorSubject<readonly T[]>([]);
+
+  private _keyManagerFactory = inject(TREE_KEY_MANAGER) as TreeKeyManagerFactory<CdkTreeNode<T, K>>;
+
+  /** The key manager for this tree. Handles focus and activation based on user keyboard input. */
+  _keyManager: TreeKeyManagerStrategy<CdkTreeNode<T, K>>;
+
+  constructor(
+    private _differs: IterableDiffers,
+    private _changeDetectorRef: ChangeDetectorRef,
+    private _dir: Directionality,
+    private _elementRef: ElementRef<HTMLElement>,
+  ) {}
 
   ngOnInit() {
     this._dataDiffer = this._differs.find([]).create(this.trackBy);
-    if (!this.treeControl && (typeof ngDevMode === 'undefined' || ngDevMode)) {
-      throw getTreeControlMissingError();
+    if (typeof ngDevMode === 'undefined' || ngDevMode) {
+      const provided = [this.treeControl, this.levelAccessor, this.childrenAccessor].filter(
+        value => !!value,
+      ).length;
+      if (provided > 1) {
+        throw getMultipleTreeControlsError();
+      } else if (provided === 0) {
+        throw getTreeControlMissingError();
+      }
```

Comment:
```suggestion
      // The tree many have either one accessor (level or children) OR a TreeControl.
      if (this.levelAccessor && this.childrenAccessor) {
        throw ...;
      } else if (this.treeControl && (this.levelAccessor || this.childrenAccessor)) {
        throw ...;
      } else if (!this.treeControl && !this.levelAccessor && !this.childrenAccessor) {
        throw ...;
      }
```

I would probably just structure this as a basic if-else block in order to give more specific error messages.

Reply:
This is what I came up with. I tried doing if/else first, but it seemed more simple to count the number of controls instead.

<img width="645" alt="image" src="https://github.com/angular/components/assets/7720245/69cce332-90ce-434c-9407-c8bce7ef2be6">

I can change it to if/else if you feel strongly. I don't feel super strongly about how this if structure, just that we should have these errors.